### DISCUSSION TITLE
How to optimize training speed for different configurations

### DISCUSSION THREAD

**Discussion by piraka9011** [NVIDIA-NeMo/NeMo#4326] [General]
First, great job to the NeMo team for the enormous effort in building such an amazing toolkit that is great for production use cases.
Not only is the code clean, but there is a significant amount of documentation that I can always refer back to, and if there's a problem, y'all are very responsive.

As someone who has been using NeMo quite often though (and scaling up usage recently), it would be great if we can compile when certain features should be used and when they shouldn't depending on the situation and use case.
For example, my understanding is that tarred manifests are great when running on multiple nodes/machines, but don't make a significant impact when running on a single node.

I'd like to start a discussion on when we should use different configs depending on our use case. This can hopefully be helpful for others coming across NeMo.

I think segmenting this by use case/setup would make the most sense, then providing different parameters and compiling them in a table of sorts (like the Riva docs https://docs.nvidia.com/deeplearning/riva/user-guide/docs/asr/asr-performance.html)

Open to suggestions of course.
Once we agree on a way of showing the data we can start compiling the data using PTL's profilers and tools with a common script.
Here's what I'm thinking:

----

# Use Cases

## Single Node, Single GPU

ex. Personal workstation with a 3090

### ModelName (ex. CitriNet-Lg)

* Dataset Format: Tar/Manfiest
* Bucketing Batch Size: int (None for no bucketing)
* Shards: int
* Num Workers: int
* Pin Memory: bool
* Accumulate Batch Size (?)
* ...what else?

## Single Node, Multiple GPUs

ex. Lab's workstation w/ 4x A100s
...

## Multiple Nodes, Single GPU

ex. AWS P3/P4s, Research HPCs
...

## Multiple Nodes, Multiple GPUs

ex. AWS P3/P4s, Research HPCs
...

**Comment by jbalam-nv** [comment]
Great suggestion @piraka9011. We are actually planning to do this over two dimensions: training data size x # GPUs. Could you let us know what the typical size of your data in hours is?

**Reply by piraka9011** [reply]
We have a "toy" dataset with a total 800 hours, but generally we use datasets of ~2500 hours (and increasing) trained either on 1 or 8 GPUs.

**Comment by titu1994** [comment]
We're doing some intensive exploration of this topic, and while that will take some time (a few months really), using some empirical evidence from the past year or so, I can give some details. 

# All configurations
- Never forget to set num_workers = 8 and pin_memory = True in **train_ds, validation_ds and test_ds (if applicable)**. We've seen slowdowns of nearly 10-20x cause of this on nearly every configuration. Recently we updated our public configs to enforce this.
  - **NOTE**: Do NOT go overboard and set a large number of workers ! It does negatively impact your training speed. We try to max out at 8 workers (even on super computers that easily have hundreds of physical cores), because every pytorch process has 8 unique number of workers (which quickly multiplies for 8 GPUs). 
- Depending on whether the single GPU is on prem (local) or virtual (GCP/AWS/etc), your scenarios are different.
- For any ASR / TTS workflow - Disable cuDNN Benchmarking. It does not work very well when shapes constantly change (time dimension in ASR and TTS).
- Try using fp16 / bf16 (if available), but check if model supports it. FP16 can be quite unstable unless your global batch size is large enough (say 1024 - 4096). Some models (Conformer) may be unstable due to algorithmic constraints (attention matrix can overflow etc).
- For RNNT training, always enable and use fused_batch_size = train_ds.batch_size // 2 or lower. You can tune this, and lower it more if OOM occurs.

Ill rank dataset sizes below on the order of "heuristics" and number of hours - don't take them as explicit rules but mere guidelines.
- Tiny datasets = D < 20 hours
- Small datasets = 20 < D < 100 hours
- Medium datasets = 100 < D < 1000 hours
- Large datasets = 1000 < D < 10000 hours
- ASR Sets = > 10000 hours

## Single GPU (Local)
- If your dataset is on an SSD, Loose file datasets are perfectly ok (no matter the size).
- If your dataset is on an HDD, buy an SSD. HDD speeds are such a huge bottleneck that training speed is severely impacted.
- No need for tarred / bucketing datasets. There is an exception to this - if your **number of files** > 1M, use tarred dataset at least. Your OS and SSD will thank you in the long run.
- Dont try to train datasets larger than Medium category (even Librispeech 960 hours on one GPU will take a few days for 100 epochs).

## Single GPU (Virtual)
- All GPU access on such instances is NOT via physical interconnect. The data is stored on a networked storage system, linked via  network mount to your VM.
- All file read and write operations over hundreds of thousands of loose files will be exceptionally slow - UNLESS you explicitly copy the network dataset onto your VM's local scratch storage (real world datasets are large enough that this is impractical, the data copy along would waste days).
- To alleviate the issues with networked IO, ALWAYS use tarred datasets, maybe even bucketing dataset (if large enough).
- Above counts for Google Colab too ! Thats why we always try to use tiny datasets there.

# Multi GPU Single Node
## Multi GPU (Local)
- Same info as Single GPU (Local)
- Use weights and biases for logging from now on. Tensorboard is just not that useful on multi GPU and WandB is so much easier to collect results on.

## Multi GPU (Virtual)
- Dataset size should be at least Medium size to consider using multi GPU on cloud, or if cost is not an issue and results are needed fast.
- Most of the info as Single GPU (Virtual) still apply.
- Take care not to overtax your CPU with large number of workers - usually on such machines the CPU is not able to process data fast enough for the GPUs.
- Reduce logging frequency to every 100 steps or more. 
- Use Weights and Biases to deal with logging.

# Multi Node Multi GPU
## Multi GPU (Slurm Cluster)
- We normally support just Slurm cluster, but recently PTL supports Horovod. We dunno much about horovod so here all advice is applicable to Slurm only.
- Try using `#SBATCH --exclusive` mode for Pytorch
- Most comments from Multi GPU (Virtual) still apply.
- Always use tarred datasets on slurm cluster. If dataset size is at ASR SET scale, use bucketing dataset.
- Models do NOT scale linearly with number of nodes. Balance the number of update steps with number of nodes, and be careful not to under train your ASR model due to large number of nodes.
